// oo7-react/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.URLBond = exports.HashBond = exports.TextBond = exports.Hash = exports.RRaisedButton = exports.Rimg = exports.Ra = exports.Rdiv = exports.Rspan = exports.ReactiveComponent = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TextField = require('material-ui/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _RaisedButton = require('material-ui/RaisedButton');

var _RaisedButton2 = _interopRequireDefault(_RaisedButton);

var _oo = require('oo7');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ReactiveComponent = exports.ReactiveComponent = function (_React$Component) {
	_inherits(ReactiveComponent, _React$Component);

	function ReactiveComponent() {
		var reactiveProps = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];
		var bonds = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

		_classCallCheck(this, ReactiveComponent);

		var _this = _possibleConstructorReturn(this, (ReactiveComponent.__proto__ || Object.getPrototypeOf(ReactiveComponent)).call(this));

		_this.reactiveProps = reactiveProps;
		_this.bonds = bonds;
		_this.allBondKeys = [].concat(reactiveProps).concat(Object.keys(bonds));
		return _this;
	}

	_createClass(ReactiveComponent, [{
		key: 'componentWillMount',
		value: function componentWillMount() {
			this.initProps();
		}
	}, {
		key: 'componentWillReceiveProps',
		value: function componentWillReceiveProps(nextProps) {
			this.updateProps(nextProps);
		}
	}, {
		key: 'componentWillUnmount',
		value: function componentWillUnmount() {
			this.finiProps();
		}
	}, {
		key: 'initProps',
		value: function initProps() {
			this.manageProps({}, this.props);
			var that = this;
			var bonds = this.bonds;
			var bondKeys = Object.keys(bonds);
			this._consolidatedExtraBonds = new _oo.ReactiveBond(bondKeys.map(function (f) {
				return bonds[f];
			}), [], function (a) {
				var s = that.state || {};
				bondKeys.forEach(function (f, i) {
					s[f] = a[i];
				});
				that.setState(s);
			}).use();
		}
	}, {
		key: 'finiProps',
		value: function finiProps() {
			if (this._consolidatedExtraBonds) {
				this._consolidatedExtraBonds.drop();
				delete this._consolidatedExtraBonds;
			}
			if (this._consolidatedBonds) {
				this._consolidatedBonds.drop();
				delete this._consolidatedBonds;
			}
		}
	}, {
		key: 'updateProps',
		value: function updateProps(nextProps) {
			this.manageProps(this.props, nextProps);
		}
	}, {
		key: 'manageProps',
		value: function manageProps(props, nextProps) {
			var that = this;
			if (this._consolidatedBonds) {
				this._consolidatedBonds.drop();
				delete this._consolidatedBonds;
			}
			this._consolidatedBonds = new _oo.ReactiveBond(this.reactiveProps.map(function (f) {
				return nextProps[f];
			}), [], function (a) {
				var s = that.state || {};
				that.reactiveProps.forEach(function (f, i) {
					s[f] = a[i];
				});
				that.setState(s);
			}).use();
		}
	}, {
		key: 'ready',
		value: function ready() {
			var _this2 = this;

			return this.allBondKeys.every(function (k) {
				return _this2.state[k] !== undefined;
			});
		}
	}, {
		key: 'readyRender',
		value: function readyRender() {
			return this.unreadyRender();
		}
	}, {
		key: 'unreadyRender',
		value: function unreadyRender() {
			return _react2.default.createElement('span', null);
		}
	}, {
		key: 'render',
		value: function render() {
			return this.ready() ? this.readyRender() : this.unreadyRender();
		}
	}]);

	return ReactiveComponent;
}(_react2.default.Component);

var Rspan = exports.Rspan = function (_ReactiveComponent) {
	_inherits(Rspan, _ReactiveComponent);

	function Rspan() {
		_classCallCheck(this, Rspan);

		return _possibleConstructorReturn(this, (Rspan.__proto__ || Object.getPrototypeOf(Rspan)).call(this, ['className', 'style', 'children']));
	}

	_createClass(Rspan, [{
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'span',
				{
					className: this.state.className,
					style: this.state.style,
					name: this.props.name
				},
				this.state.children
			);
		}
	}]);

	return Rspan;
}(ReactiveComponent);

var Rdiv = exports.Rdiv = function (_ReactiveComponent2) {
	_inherits(Rdiv, _ReactiveComponent2);

	function Rdiv() {
		_classCallCheck(this, Rdiv);

		return _possibleConstructorReturn(this, (Rdiv.__proto__ || Object.getPrototypeOf(Rdiv)).call(this, ['className', 'style', 'children']));
	}

	_createClass(Rdiv, [{
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'div',
				{
					className: this.state.className,
					style: this.state.style,
					name: this.props.name
				},
				this.state.children
			);
		}
	}]);

	return Rdiv;
}(ReactiveComponent);

var Ra = exports.Ra = function (_ReactiveComponent3) {
	_inherits(Ra, _ReactiveComponent3);

	function Ra() {
		_classCallCheck(this, Ra);

		return _possibleConstructorReturn(this, (Ra.__proto__ || Object.getPrototypeOf(Ra)).call(this, ['href', 'target', 'className', 'style', 'children']));
	}

	_createClass(Ra, [{
		key: 'render',
		value: function render() {
			return _react2.default.createElement(
				'a',
				{
					href: this.state.href,
					target: this.state.target,
					className: this.state.className,
					style: this.state.style,
					name: this.props.name
				},
				this.state.children
			);
		}
	}]);

	return Ra;
}(ReactiveComponent);

var Rimg = exports.Rimg = function (_ReactiveComponent4) {
	_inherits(Rimg, _ReactiveComponent4);

	function Rimg() {
		_classCallCheck(this, Rimg);

		return _possibleConstructorReturn(this, (Rimg.__proto__ || Object.getPrototypeOf(Rimg)).call(this, ['src', 'className', 'style']));
	}

	_createClass(Rimg, [{
		key: 'render',
		value: function render() {
			return _react2.default.createElement('img', {
				src: this.state.src,
				className: this.state.className,
				style: this.state.style,
				name: this.props.name
			});
		}
	}]);

	return Rimg;
}(ReactiveComponent);

var RRaisedButton = exports.RRaisedButton = function (_ReactiveComponent5) {
	_inherits(RRaisedButton, _ReactiveComponent5);

	function RRaisedButton() {
		_classCallCheck(this, RRaisedButton);

		return _possibleConstructorReturn(this, (RRaisedButton.__proto__ || Object.getPrototypeOf(RRaisedButton)).call(this, ['disabled', 'label']));
	}

	_createClass(RRaisedButton, [{
		key: 'render',
		value: function render() {
			return _react2.default.createElement(_RaisedButton2.default, {
				disabled: this.state.disabled,
				label: this.state.label,
				onClick: this.props.onClick,
				name: this.props.name
			});
		}
	}]);

	return RRaisedButton;
}(ReactiveComponent);

var Hash = exports.Hash = function (_ReactiveComponent6) {
	_inherits(Hash, _ReactiveComponent6);

	function Hash() {
		_classCallCheck(this, Hash);

		return _possibleConstructorReturn(this, (Hash.__proto__ || Object.getPrototypeOf(Hash)).call(this, ['value', 'className', 'style']));
	}

	_createClass(Hash, [{
		key: 'render',
		value: function render() {
			var v = this.state.value;
			var d = typeof v === 'string' && v.startsWith('0x') && v.length >= 18 ? v.substr(0, 8) + '…' + v.substr(v.length - 4) : v;
			return _react2.default.createElement(
				'span',
				{
					className: this.state.className,
					style: this.state.style,
					title: this.state.value,
					name: this.props.name
				},
				d
			);
		}
	}]);

	return Hash;
}(ReactiveComponent);

Hash.defaultProps = {
	className: '_hash'
};

var TextBond = exports.TextBond = function (_React$Component2) {
	_inherits(TextBond, _React$Component2);

	function TextBond() {
		_classCallCheck(this, TextBond);

		var _this9 = _possibleConstructorReturn(this, (TextBond.__proto__ || Object.getPrototypeOf(TextBond)).call(this));

		_this9.state = { value: '', validity: false };
		return _this9;
	}

	_createClass(TextBond, [{
		key: 'fixValue',
		value: function fixValue(v) {
			var f = function (b) {
				this.setState({ value: v, validity: b });
				if (this.props.bond instanceof _oo.Bond) {
					if (b) this.props.bond.changed(v);else this.props.bond.reset();
				}
			}.bind(this);

			this.setState({ value: v, validity: null });

			if (typeof this.props.validator !== 'function') {
				f(true);
			} else {
				var a = v !== undefined && this.props.validator(v);
				if (a instanceof Promise || a instanceof _oo.Bond) {
					a.then(f);
				} else {
					f(a);
				}
			}
		}
	}, {
		key: 'componentWillMount',
		value: function componentWillMount() {
			this.fixValue(this.state.value);
		}
	}, {
		key: 'render',
		value: function render() {
			var _this10 = this;

			return _react2.default.createElement(_TextField2.default, {
				className: this.props.className,
				style: this.props.style,
				name: this.props.name,
				value: this.state.value,
				floatingLabelText: this.props.floatingLabelText,
				errorText: this.props.errorText || (!this.state.validity ? this.props.invalidText : null),
				onChange: function onChange(e, v) {
					return _this10.fixValue(v);
				}
			});
		}
	}]);

	return TextBond;
}(_react2.default.Component);

TextBond.defaultProps = {
	invalidText: 'Invalid'
};

var HashBond = exports.HashBond = function (_TextBond) {
	_inherits(HashBond, _TextBond);

	function HashBond() {
		_classCallCheck(this, HashBond);

		return _possibleConstructorReturn(this, (HashBond.__proto__ || Object.getPrototypeOf(HashBond)).apply(this, arguments));
	}

	return HashBond;
}(TextBond);

HashBond.defaultProps = {
	floatingLabelText: 'Enter a hash to look up',
	invalidText: 'Invalid 32-byte hash',
	validator: function validator(v) {
		return v.startsWith('0x') && v.length == 66;
	}
};

var URLBond = exports.URLBond = function (_TextBond2) {
	_inherits(URLBond, _TextBond2);

	function URLBond() {
		_classCallCheck(this, URLBond);

		return _possibleConstructorReturn(this, (URLBond.__proto__ || Object.getPrototypeOf(URLBond)).apply(this, arguments));
	}

	return URLBond;
}(TextBond);

URLBond.defaultProps = {
	floatingLabelText: 'Enter a URL',
	invalidText: 'Not a URL',
	validator: function validator(u) {
		try {
			return new URL(u) && true;
		} catch (e) {
			return false;
		}
	}
};
