// parity-reactive-ui/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.TransactionProgressBadge = exports.AddressBond = exports.BalanceInput = exports.SignatureProgress = exports.TransactionProgress = exports.RichAccount = exports.Account = exports.AccountIcon = exports.BlockNumber = exports.Balance = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _bignumber = require('bignumber.js');

var _bignumber2 = _interopRequireDefault(_bignumber);

var _blockies = require('blockies');

var _blockies2 = _interopRequireDefault(_blockies);

var _oo = require('oo7');

var _oo7Parity = require('oo7-parity');

var _oo7React = require('oo7-react');

var _TransactionProgressBadge = require('./TransactionProgressBadge.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function createIdentityImage(address) {
	var scale = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 8;

	return (0, _blockies2.default)({
		seed: (address || '').toLowerCase(),
		size: 8,
		scale: scale
	}).toDataURL();
}

var Balance = exports.Balance = function (_ReactiveComponent) {
	_inherits(Balance, _ReactiveComponent);

	function Balance() {
		_classCallCheck(this, Balance);

		return _possibleConstructorReturn(this, (Balance.__proto__ || Object.getPrototypeOf(Balance)).call(this, ['value']));
	}

	_createClass(Balance, [{
		key: 'render',
		value: function render() {
			if (this.state.value === null || typeof this.state.value == 'undefined') return _react2.default.createElement(
				'span',
				{ className: '_undefined _balance' },
				'?'
			);
			var v = new _bignumber2.default(this.state.value);
			var isNeg = v.lt(0);
			var s = (0, _oo7Parity.splitValue)(v.mul(isNeg ? -1 : 1));
			var a = ('' + s.base.mul(1000).round().div(1000)).replace(/(\d)(?=(\d{3})+$)/g, "$1,");
			return _react2.default.createElement(
				'span',
				{ className: '_balance _' + _oo7Parity.denominations[s.denom] },
				isNeg ? "-" : this.props.signed ? "+" : "",
				a,
				_react2.default.createElement(
					'span',
					{ className: '_denom' },
					_oo7Parity.denominations[s.denom]
				)
			);
		}
	}]);

	return Balance;
}(_oo7React.ReactiveComponent);

var BlockNumber = exports.BlockNumber = function (_ReactiveComponent2) {
	_inherits(BlockNumber, _ReactiveComponent2);

	function BlockNumber() {
		_classCallCheck(this, BlockNumber);

		return _possibleConstructorReturn(this, (BlockNumber.__proto__ || Object.getPrototypeOf(BlockNumber)).call(this, ['value']));
	}

	_createClass(BlockNumber, [{
		key: 'render',
		value: function render() {
			var classes = this.props.classes === null ? '_blocknumber' : this.props.classes;
			var undefClasses = this.props.undefClasses === null ? '_blocknumber _undefined' : this.props.undefClasses;
			var undefContent = this.props.undefContent === null ? '?' : this.props.undefContent;
			if (this.state.value === null || typeof this.state.value == 'undefined') return _react2.default.createElement(
				'span',
				{ className: undefClasses },
				undefContent
			);
			var a = ('' + this.state.value).replace(/(\d)(?=(\d{3})+$)/g, "$1,");
			return _react2.default.createElement(
				'span',
				{ className: classes },
				'#',
				a
			);
		}
	}]);

	return BlockNumber;
}(_oo7React.ReactiveComponent);

;

var AccountIcon = exports.AccountIcon = function (_ReactiveComponent3) {
	_inherits(AccountIcon, _ReactiveComponent3);

	function AccountIcon() {
		_classCallCheck(this, AccountIcon);

		return _possibleConstructorReturn(this, (AccountIcon.__proto__ || Object.getPrototypeOf(AccountIcon)).call(this, ['address', 'className', 'style']));
	}

	_createClass(AccountIcon, [{
		key: 'render',
		value: function render() {
			if (typeof this.state.address == "string") {
				return _react2.default.createElement('img', {
					src: createIdentityImage(this.state.address),
					style: this.state.style,
					className: typeof this.state.className === 'string' ? this.state.className : '',
					id: this.props.id,
					'data-address-img': true
				});
			} else {
				return _react2.default.createElement(
					'span',
					{
						style: this.props.undefStyle,
						className: this.props.undefClassName
					},
					this.props.undefContent
				);
			}
		}
	}]);

	return AccountIcon;
}(_oo7React.ReactiveComponent);

;
AccountIcon.defaultProps = {
	style: {},
	className: '_accountIcon',
	undefStyle: {},
	undefClassName: '_accountIcon _undefined',
	undefContent: '',
	id: null
};

var Account = exports.Account = function (_ReactiveComponent4) {
	_inherits(Account, _ReactiveComponent4);

	function Account() {
		_classCallCheck(this, Account);

		return _possibleConstructorReturn(this, (Account.__proto__ || Object.getPrototypeOf(Account)).call(this, ['address'], { accountsInfo: parity.bonds.accountsInfo }));
	}

	_createClass(Account, [{
		key: 'render',
		value: function render() {
			if (typeof this.state.address == "string") {
				var i = this.state.accountsInfo != null ? this.state.accountsInfo[parity.api.util.toChecksumAddress(this.state.address)] : null;
				var a = i == null ? this.state.address.substr(0, 8) + "…" + this.state.address.substr(38) : i.name;
				return _react2.default.createElement(
					'span',
					{ className: '_account' },
					this.props.icon ? _react2.default.createElement('img', { src: createIdentityImage(this.state.address), className: '_identigram', 'data-address-img': true, style: { marginRight: '0.5ex' } }) : '',
					this.props.text ? a : ''
				);
			} else {
				return _react2.default.createElement(
					'span',
					{ className: 'undefined _account' },
					'[null]'
				);
			}
		}
	}]);

	return Account;
}(_oo7React.ReactiveComponent);

;
Account.defaultProps = {
	icon: true,
	text: true
};

var RichAccount = exports.RichAccount = function (_ReactiveComponent5) {
	_inherits(RichAccount, _ReactiveComponent5);

	function RichAccount() {
		_classCallCheck(this, RichAccount);

		return _possibleConstructorReturn(this, (RichAccount.__proto__ || Object.getPrototypeOf(RichAccount)).call(this, ['address']));
	}

	_createClass(RichAccount, [{
		key: 'render',
		value: function render() {
			if (typeof this.state.address == "string") {
				return _react2.default.createElement(
					'span',
					null,
					_react2.default.createElement(Account, { address: this.state.address }),
					' ',
					_react2.default.createElement(Balance, { value: parity.api.eth.getBalance.blockBond(this.state.address) })
				);
			} else {
				return _react2.default.createElement(
					'span',
					{ className: 'undefined _account' },
					'[null]'
				);
			}
		}
	}]);

	return RichAccount;
}(_oo7React.ReactiveComponent);

;

var TransactionProgress = exports.TransactionProgress = function (_ReactiveComponent6) {
	_inherits(TransactionProgress, _ReactiveComponent6);

	function TransactionProgress() {
		_classCallCheck(this, TransactionProgress);

		return _possibleConstructorReturn(this, (TransactionProgress.__proto__ || Object.getPrototypeOf(TransactionProgress)).call(this, ['request']));
	}

	_createClass(TransactionProgress, [{
		key: 'render',
		value: function render() {
			if (_typeof(this.state.request) != 'object' || this.state.request == null) return _react2.default.createElement('div', { className: '_progress _null' });
			var x;
			if (x = this.state.request.requested) return _react2.default.createElement(
				'div',
				{ className: '_progress _authorising' },
				'Authorising transaction...'
			);
			if (x = this.state.request.signed) return _react2.default.createElement(
				'div',
				{ className: '_progress _submitting' },
				'Confirming transaction...'
			);
			if (x = this.state.request.confirmed) return _react2.default.createElement(
				'div',
				{ className: '_progress _confirmed' },
				'Confirmed at ',
				_react2.default.createElement(BlockNumber, { value: x.blockNumber })
			);
			if (x = this.state.request.failed) return _react2.default.createElement(
				'div',
				{ className: '_progress _failed' },
				'Failed: ',
				x.text
			);
			return _react2.default.createElement(
				'div',
				null,
				'???'
			);
		}
	}]);

	return TransactionProgress;
}(_oo7React.ReactiveComponent);

var SignatureProgress = exports.SignatureProgress = function (_ReactiveComponent7) {
	_inherits(SignatureProgress, _ReactiveComponent7);

	function SignatureProgress() {
		_classCallCheck(this, SignatureProgress);

		return _possibleConstructorReturn(this, (SignatureProgress.__proto__ || Object.getPrototypeOf(SignatureProgress)).call(this, ['request']));
	}

	_createClass(SignatureProgress, [{
		key: 'render',
		value: function render() {
			if (_typeof(this.state.request) != 'object' || this.state.request == null) return _react2.default.createElement('div', { className: '_progress _null' });
			var x;
			if (x = this.state.request.requested) return _react2.default.createElement(
				'div',
				{ className: '_progress _authorising' },
				'Authorising signature...'
			);
			if (x = this.state.request.signed) return _react2.default.createElement(
				'div',
				{ className: '_progress _authorised' },
				'Authorised'
			);
			if (x = this.state.request.failed) return _react2.default.createElement(
				'div',
				{ className: '_progress _failed' },
				'Failed: ',
				x.text
			);
			return _react2.default.createElement('div', null);
		}
	}]);

	return SignatureProgress;
}(_oo7React.ReactiveComponent);

var BalanceInput = exports.BalanceInput = function (_React$Component) {
	_inherits(BalanceInput, _React$Component);

	function BalanceInput() {
		_classCallCheck(this, BalanceInput);

		var _this8 = _possibleConstructorReturn(this, (BalanceInput.__proto__ || Object.getPrototypeOf(BalanceInput)).call(this));

		_this8.state = { value: 1 };
		return _this8;
	}

	return BalanceInput;
}(_react2.default.Component);

var AddressBond = exports.AddressBond = function (_TextBond) {
	_inherits(AddressBond, _TextBond);

	function AddressBond() {
		_classCallCheck(this, AddressBond);

		return _possibleConstructorReturn(this, (AddressBond.__proto__ || Object.getPrototypeOf(AddressBond)).apply(this, arguments));
	}

	return AddressBond;
}(_oo7React.TextBond);

AddressBond.defaultProps = {
	floatingLabelText: 'Enter an address',
	invalidText: 'Invalid 20-byte address',
	validator: function validator(v) {
		return v.startsWith('0x') && v.length == 42;
	}
};

exports.TransactionProgressBadge = _TransactionProgressBadge.TransactionProgressBadge;
